### PR TITLE
Added body expression rule

### PR DESCRIPTION
**Pull request by aballano**
Hope @vanniktech could benefit from this :)

Let me know if this is the right format for the test as I've seen different ways already.

### CODE REVIEW

**Comment by aballano** [comment]
Well, it seems to work :D Should I fix the issues? Anyway is weird because it should be disabled by default, right?

<details>

> Ruleset: style
> 	BodyExpression - [toString] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt:21:3
> 	BodyExpression - [valueOrDefaultInternal] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt:49:3
> 	BodyExpression - [toString] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Issue.kt:16:3
> 	BodyExpression - [toString] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Issue.kt:47:3
> 	BodyExpression - [startLineAndColumn] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Location.kt:35:4
> 	BodyExpression - [searchName] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Signatures.kt:14:2
> 	BodyExpression - [searchSignature] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Signatures.kt:42:2
> 	BodyExpression - [toString] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/YamlConfig.kt:29:3
> 	BodyExpression - [convert] at detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/PathConverters.kt:25:3
> 	BodyExpression - [toString] at detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/baseline/Baseline.kt:9:3
> 	BodyExpression - [toString] at detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/baseline/Blacklist.kt:13:3
> 	BodyExpression - [toString] at detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/baseline/Whitelist.kt:13:3
> 	BodyExpression - [render] at detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReport.kt:36:4
> 	BodyExpression - [render] at detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReport.kt:39:4
> 	BodyExpression - [extractRuleToRuleSetIdMap] at detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReport.kt:44:3
> 	BodyExpression - [cannotGenerate] at detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/ComplexityReportGenerator.kt:36:3
> 	BodyExpression - [instance] at detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/DetektFacade.kt:18:3
> 	BodyExpression - [instance] at detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/DetektFacade.kt:22:3
> 	BodyExpression - [instance] at detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/DetektFacade.kt:28:3
> 	BodyExpression - [determineLineSeparator] at detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtCompiler.kt:42:4
> 	BodyExpression - [runAsync] at detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Tasks.kt:20:2
> 	BodyExpression - [task] at detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Tasks.kt:24:2
> 	BodyExpression - [count] at detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/processors/ProjectSLOCProcessor.kt:26:4
> 	BodyExpression - [isEscaped] at detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/processors/util/LLOC.kt:80:4
> 	BodyExpression - [frequency] at detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/processors/util/LLOC.kt:86:5
> 	BodyExpression - [<anonymous>] at detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/ExpressionBodySyntax.kt:48:5
> 	BodyExpression - [instance] at detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt:15:3
> 	BodyExpression - [profileArgumentsOrDefault] at detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektExtension.kt:32:3
> 	BodyExpression - [arguments] at detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/ProfileExtension.kt:20:3
> 	BodyExpression - [instance] at detekt-migration/src/main/kotlin/io/gitlab/arturbosch/detekt/migration/MigrationRuleSetProvider.kt:15:3
> 	BodyExpression - [calc] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/FeatureEnvy.kt:98:5
> 	BodyExpression - [isPublic] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/Junk.kt:24:2
> 	BodyExpression - [hasInvalidLoopRange] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt:37:4
> 	BodyExpression - [getIntValueForElement] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt:48:3
> 	BodyExpression - [callsNextMethod] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/IteratorHasNextCallsNextMethod.kt:37:3
> 	BodyExpression - [visitClass] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/WrongEqualsTypeParameter.kt:20:4
> 	BodyExpression - [throwsNoSuchElementExceptionThrown] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/iterator/IteratorExtension.kt:21:2
> 	BodyExpression - [frequency] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/ComplexCondition.kt:63:4
> 	BodyExpression - [getLiteralsOverThreshold] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/StringLiteralDuplication.kt:43:4
> 	BodyExpression - [visitObjectDeclaration] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:42:4
> 	BodyExpression - [hasPublicVisibility] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/empty/EmptyDefaultConstructor.kt:22:3
> 	BodyExpression - [isPotentialMethod] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ExceptionRaisedInUnexpectedLocation.kt:38:3
> 	BodyExpression - [hasThrowExpression] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ExceptionRaisedInUnexpectedLocation.kt:42:3
> 	BodyExpression - [isNotInInnerFunction] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ReturnFromFinally.kt:31:3
> 	BodyExpression - [hasParameterReferences] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/SwallowedException.kt:38:3
> 	BodyExpression - [callsMemberOfCaughtException] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/SwallowedException.kt:44:3
> 	BodyExpression - [isMainFunction] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ThrowingExceptionInMain.kt:28:3
> 	BodyExpression - [hasArgsParameter] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ThrowingExceptionInMain.kt:32:3
> 	BodyExpression - [isStringArrayParameter] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ThrowingExceptionInMain.kt:36:3
> 	BodyExpression - [containsThrowExpression] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ThrowingExceptionInMain.kt:40:3
> 	BodyExpression - [createsSameExceptionType] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ThrowingNewInstanceOfSameException.kt:35:3
> 	BodyExpression - [hasSameExceptionParameter] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ThrowingNewInstanceOfSameException.kt:39:3
> 	BodyExpression - [<anonymous>] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/performance/ForEachOnRange.kt:23:5
> 	BodyExpression - [instance] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/CodeSmellProvider.kt:15:3
> 	BodyExpression - [instance] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/CommentSmellProvider.kt:19:3
> 	BodyExpression - [instance] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/ComplexityProvider.kt:24:3
> 	BodyExpression - [instance] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/EmptyCodeProvider.kt:16:3
> 	BodyExpression - [instance] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/ExceptionsProvider.kt:26:3
> 	BodyExpression - [instance] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/PerformanceProvider.kt:15:3
> 	BodyExpression - [instance] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/PotentialBugProvider.kt:29:3
> 	BodyExpression - [instance] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/StyleGuideProvider.kt:38:3
> 	BodyExpression - [filterPackageStatements] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/FileParsingRule.kt:59:4
> 	BodyExpression - [filterImportStatements] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/FileParsingRule.kt:66:4
> 	BodyExpression - [visitConstantExpression] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt:52:4
> 	BodyExpression - [normalizeForParsingAsDouble] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt:101:3
> 	BodyExpression - [indexOfFirstMember] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt:52:4
> 	BodyExpression - [visitClass] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseDataClass.kt:31:4
> 	BodyExpression - [hasOnlyUtilityClassMembers] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UtilityClassWithPublicConstructor.kt:37:3
> 	BodyExpression - [isCompanionObject] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UtilityClassWithPublicConstructor.kt:43:3
> 	BodyExpression - [hasPublicConstructor] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UtilityClassWithPublicConstructor.kt:50:4
> 	BodyExpression - [<anonymous>] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/WildcardImport.kt:32:5
> 	BodyExpression - [<anonymous>] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/WildcardImport.kt:36:5
> 	BodyExpression - [visitNamedDeclaration] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/naming/NamingRules.kt:62:4
> 	BodyExpression - [withinObjectDeclaration] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/naming/NamingRules.kt:108:3
> 	BodyExpression - [countBreakAndReturnStatements] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatements.kt:31:3
> 	BodyExpression - [visitKtElement] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatements.kt:39:6
> 	BodyExpression - [instance] at detekt-sample-extensions/src/main/kotlin/io/gitlab/arturbosch/detekt/sample/extensions/SampleProvider.kt:17:3
> 	BodyExpression - [render] at detekt-sample-extensions/src/main/kotlin/io/gitlab/arturbosch/detekt/sample/extensions/reports/QualifiedNamesConsoleReport.kt:12:3
> 	BodyExpression - [render] at detekt-sample-extensions/src/main/kotlin/io/gitlab/arturbosch/detekt/sample/extensions/reports/QualifiedNamesOutputReport.kt:15:3

</details

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/BodyExpressionSpec.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * @author aballano
+ */
+class BodyExpressionSpec : SubjectSpek<BodyExpression>({
+
+	subject { BodyExpression() }
+
+	describe("a simple test") {
+
+		it("should find functions with return statements that can be body expressions") {
```
Comment (arturbosch) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/BodyExpressionSpec.kt]: Looks good to me (basically we already have this kind of rule inside the formatting package :D (which gets removed)).

As your going the way from `ReturnStatement` -> `KtBlockStatement` you should include test cases for other blocks like lambdas, inits, constructors etc so we do not have false positives.

**Comment by arturbosch** [comment]
Indeed its wierd ... do the code smells get reported locally?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/BodyExpressionSpec.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+			val findings = subject.lint(code)
+			Assertions.assertThat(findings).hasSize(2)
+		}
+	}
+
+})
+
+private val code: String =
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/BodyExpressionSpec.kt]: Please move this code to it's own class in the *cases* package.

File: detekt-cli/src/main/resources/default-detekt-config.yml [hunk 3]
```diff
@@ -60,6 +60,8 @@ potential-bugs:
     active: false
   UselessPostfixExpression:
     active: false
+  ExpressionBody:
```
Comment (schalkms) [inline detekt-cli/src/main/resources/default-detekt-config.yml]: Your class is named *BodyExpression*. Thus, this declaration is wrong.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/BodyExpressionSpec.kt (hunk 1 above)
Comment (aballano) [reply detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/BodyExpressionSpec.kt]: So is this actually done by ExpressionBodySyntax? Then there's no point on adding this rule, is it?

Comment (arturbosch) [reply detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/BodyExpressionSpec.kt]: No not really, sorry :/. I will port the ExpressionBodySyntax rule to the default rule set.

**Comment by arturbosch** [comment]
I will close this, see other comment. But thanks for finding out that the isActive check is kinda broken. We have another issue now for that!

**Pull request closed without merging**